### ISSUE TITLE
Sparse construction uses dense logic

### ISSUE THREAD

**Issue by alecjacobson** [pressureless/linear_algebra#45]
The output (e.g., python) code for the tutorial example

```I❤️LA
given
n ∈ ℤ
E ∈ {ℤ²}

A_i,j = { 1 if (i,j) ∈ E
          0 otherwise

where
A ∈ ℝ^(n × n)
```

Will output correct code, but the logic assumes a dense graph (double for loop consider every element). 

```
    for i in range(1, n+1):
        for j in range(1, n+1):
            if (i, j) in E:
```

I'm not 100% sure, but I think this is O(|E| n^2) which assuming no duplicates in E becomes O(n^3).

Shouldn't this be something more like:

```
    for (i, j) in E:
```
 ?

**Comment by pressureless** [comment]
Yes. The current output is not efficient for the example. It should be optimized.

I just used the `two loops` for all `if conditions`, the following example may look better.
`
given
n ∈ ℤ
E ∈ {ℤ²}

A_i,j = { 1   if i + j < 10
              0  otherwise

where
A ∈ ℝ^(n × n)
`
